docs/design-proposals/node/Unified-Vertical-Scaling-for-Container-and-VM-Pods.md

```diff
@@ -0,0 +1,646 @@
+   introduction and comparision between them.
+   
+   Memory Ballooning is primarily used by the hypervisor to reclaim some "unused" memory blocks from the guest OS. Combined
+   with hypervisor logic and the "ballooning driver" added into the guest OS, some unused memory blocks can be reclaimed from
+   a guest OS and the hypervisor can reused them to other VMs that require more memory allocation. 
+   
+   Memory Ballooning provides a good way for supporting memory over-subscription scenarios. Because the "free" memory is tracked
+   by the ballooning driver before the memory reclaim request from the hypervisor, they can be reclaimed fairly quick when 
+   more memory is needed at the host.
+   
+   However,  using Memory Ballooning supporting vertical scaling feature, there are few hard limitations:
+   
+     1. Memory Ballooning is limited between the maxBootTimeMemoryAllocation and the currentMemoryAllocation defined in the 
+        domain with hard cap for memory scale-up.
+        The maxBootTimeMemoryAllocation cannot be reset without restarting the VM. For example, given a VM with
+        requested and limit of memory as 4G, essentially there is no space to scale up the VM with ballooning. The current 
+        design would have to be changed to either increase the maxBootTimeMemoryAllocation or lower the currentMemoryAllocation
+        to have ballooning to work. Even though, either way will not solve this limitation. Increasing maxBootTimeMemoryAllocation
+        too high can introduce some unpredicted behavior of the VM ( and host as well) when it is restarted since it will 
+        use this setting to boot the VM up before ballooning driver kicks in to reset the memory allocation to the 
+        currentMemoryAllocation level. Some changes might be made into the hypervisor layer to address this situation. 
+        However it will take quite some logic to achieve this goal. For example, how to differentiate the intention and 
+        boot the VM with different settings( future work [2]).
+        
+        One possible way to mitigate this limitation is to control the scale up of a VM by some policies so that the 
+        MaxBootTimeMemoryAllocation can be controlled to some level. however, this is not a reliable solution and has to be
+        coordinated with other features such as number of VMs scheduled on the host, level of over-subscription on the host etc.
+      
+     2. Memory fragmentation
+        The nature of the ballooning mechanism is to reclaim unused memory blocks of the guest OS and render them to host 
+        hypervisor. Given time and with repeated memory increase and decrease, it can left the VM memory fragmentation 
+        and affect its performance.
+      
+     3. Ballooning is not NUMA aware
+
+   Memory hotplug/unplug simulates the physical memory device being plugged into the physical machines for a guest OS. 
```

Comment:
This paragraph requires indent. Also it focuses mostly on something else instead of NUMA.

Reply:
This paragraph is not about the #3 above. i thought #3 is self-explained, so there is no explanation under it. i reworded the #3 and this paragraph to avoid such impressions.